### ISSUE TITLE
Geometric aggregate takes too long on point in polygon intersection

### ISSUE THREAD

**Issue by tschmetzer** [qgis/QGIS#44463]
### What is the bug about?

An aggregate expression in the field calculator shows very slow behaviour. It almost seems to hang but will probably finish after hours. The situation occurs when using an aggregate expression with a location filter by geometry. I guess this is just due to a not optimised implementation of the aggregate function that I use on a polygon layer:

```
aggregate(
layer:= '<point_layer>',
aggregate:='sum',
expression:=dni,
filter:=intersects($geometry, geometry(@parent))
)
```

There are phenomenas with this expression:
- With a large **indexed** point layer file like [DNI_lowres_sunbelt_pixels2points.sqlite](https://drive.google.com/file/d/1hK6qtymSjb32JiMR-5EnxqUbUMeER2xF/view?usp=sharing) (watchout **this file is yet unindexed** and has 4.5 GB and needs to be converted to Geopackage or SpatiaLite format prior to use) the expression editor dialog hangs as soon as the layer name is introduced. No editing of the expression is possible anymore after that point.
- With a small point layer file like [DNI_lowres_sunbelt_pixels2points_tinypart.zip](https://github.com/qgis/QGIS/files/6906698/DNI_lowres_sunbelt_pixels2points_tinypart.zip) the calculation took about 25 minutes in my PC for an evaluation of 5,000 points (150kb file size) in 3 polygons among 12,606 polygons (5 MB file size).

The polygon layer I used: [Beck_KG_V1_present_0p083_sunbelt_polygonized.zip](https://github.com/qgis/QGIS/files/6907103/Beck_KG_V1_present_0p083_sunbelt_polygonized.zip)

### Steps to reproduce the issue

1. Open Field calculator on layer Beck_KG_V1_present_0p083_sunbelt_polygonized
2. Insert aggregate expression adapted from above:
  ```
  aggregate(
  layer:= 'DNI_lowres_sunbelt_pixels2points(_tinypart)',
  aggregate:='sum',
  expression:=dni,
  filter:=intersects($geometry, geometry(@parent))
  )
  ```

![grafik](https://user-images.githubusercontent.com/67278094/127635398-40e636e4-01e5-4821-90e8-ba1f46548eff.png)

3. (Re-)Place the correct layer name/id into the aggregate expression
  --> With a large point layer QGIS hangs here already.

3. Hit "OK". 
  --> With a small point layer this takes very long.

### Versions

QGIS version | 3.20.1-Odense | QGIS code revision | 1c3c5cd6
-- | -- | -- | --
Qt version | 5.15.2
Python version | 3.9.5
GDAL/OGR version | 3.3.1
PROJ version | 8.1.0
EPSG Registry database version | v10.027 (2021-06-17)
GEOS version | 3.9.1-CAPI-1.14.2
SQLite version | 3.35.2
PDAL version | 2.3.0
PostgreSQL client version | 13.0
SpatiaLite version | 5.0.1
QWT version | 6.1.3
QScintilla2 version | 2.11.5
OS version | Windows 10 Version 2009
  |   |   |  
Active Python plugins | GroupStatslatlontoolsnominatimQuickOSMdb_managerprocessing

**Comment by gioman** [comment]
> [DNI_lowres_sunbelt_pixels2points_tinypart.zip](https://github.com/qgis/QGIS/files/6906698/DNI_lowres_sunbelt_pixels2points_tinypart.zip)

@tschmetzer it only contains the points, please also add the polygons.

**Comment by tschmetzer** [comment]
> 
> 
> > [DNI_lowres_sunbelt_pixels2points_tinypart.zip](https://github.com/qgis/QGIS/files/6906698/DNI_lowres_sunbelt_pixels2points_tinypart.zip)
> 
> @tschmetzer it only contains the points, please also add the polygons.

[Draft] to be continued in an hour or so.. ;)

**Comment by roya0045** [comment]
Not sure how this is a bug if you are constantly requesting expensive expressions instead of doing a static join once. To fix the bug, reduce the dataset or don't do it this way, IMO. There is no magic trick with this.

**Comment by tschmetzer** [comment]
> 
> 
> Not sure how this is a bug if you are constantly requesting expensive expressions instead of doing a static join once. To fix the bug, reduce the dataset or don't do it this way, IMO. There is no magic trick with this.

I understand the expression takes so long because it seems to be expensive. Can someone elaborate a bit more why the expression is so expensive compared to a static join? Could it be improved reordering database requests?

And why does the editor hang/freeze?  I guess it's due to the expensive evaluation of a first example expression. Can the evaluation of only one example really take forever?

**Comment by gioman** [comment]
> And why does the editor hang/freeze?

@tschmetzer can you attach the polygons necessary to test your expression with your points?

**Comment by tschmetzer** [comment]
> > And why does the editor hang/freeze?
> 
> @tschmetzer can you attach the polygons necessary to test your expression with your points?

I don't understand what polygons you are referring to. Maybe there's a misunderstandig. 
Are you referring to the three polygons that actually match with the points? Do you want me to isolate them and store them in a separate layer file that I upload here?

**Comment by roya0045** [comment]
> I understand the expression takes so long because it seems to be expensive. Can someone elaborate a bit more why the expression is so expensive compared to a static join? Could it be improved reordering database requests?
> 
> And why does the editor hang/freeze? I guess it's due to the expensive evaluation of a first example expression. Can the evaluation of only one example really take forever?

1) if it takes you 20 minutes to calculate something and your colleague always ask you the same question, is it simpler to have the answer on hand or to redo the calculations each time? I'm also guessing the results are not cached due to slight changes each time.

2) If your map render depend on this expression, then the expression must be done before rendering, there is no magic to this.

**Comment by agiudiceandrea** [comment]
@tschmetzer your Beck_KG_V1_present_0p083_sunbelt_polygonized.shp layer contains 12.606, so the same expressions will be executed for 12.606 times and each time an aggregate function between a layer with 12.606 features and a layer with 5.000 features will be executed.

I would try to narrow down the times the aggregate function will actually need to be executed just adding a bounding box intersection condition:

if(intersects(bounds($geometry),layer_property('DNI_lowres_sunbelt_pixels2points_tinypart','extent')),
aggregate(
layer:= 'DNI_lowres_sunbelt_pixels2points_tinypart',
aggregate:='sum',
expression:=dni,
filter:=intersects($geometry, geometry(@parent))
),0)

**Comment by tschmetzer** [comment]
@agiudiceandrea Thanks for that great explanation!

> I would try to narrow down the times the aggregate function will actually need to be executed just adding a bounding box intersection condition:
> [...]

That looks like a splendid approach!

Just as an idea to improve QGIS for all users without the knowlegde what operations will be executed in detail: Is there a way to integrate this approach into all geometric filters for expressions or the aggregate function? I am not experienced enough to jugde where geometric filter expressions could be used or what such an optimization could cause. Is there anything that contradicts integrating such an optimization?
Or if there are doubts about just jeopardizing the purity of the function - what about providing an optimized version of the filter expression or aggregate function?

**Comment by tschmetzer** [comment]
I am not convinced of closing this issue at the current state:
- The field calculator shouldn't hang when inserting an expression even for large point layers (see step 3)
- What are the arguments against a feature request for adding a bounding box intersection condition at least into an *optimised intersection filter* according to @agiudiceandrea's proposal ?

**Comment by gioman** [comment]
>     * against

I guess none, so please file the feature request.

**Comment by tschmetzer** [comment]
Independant of consequence the expression will have when repeatedly used on many features I am still unhappy with the hang in the field calculator. Is this presumably because all features are considered in order to evaluate the expression for the current feature or because there is a previewing (pre-)calculation for other features that might come next?

**Comment by tschmetzer** [comment]
> Independant of the consequence the expression will have when repeatedly used on many features I am still unhappy with the hang in the field calculator. Is this presumably because all features are considered in order to evaluate the expression for the current feature or because there is a previewing (pre-)calculation for other features that might come next?

Comments appreciated :)

**Comment by Jeroen-GroeneBij** [comment]
I have found similar behaviour in using aggregate in Attribute form to get a value form a field of the polygon layer my point is in, using the same function as TS.
Writing the expression immediately lead to slow response. Also left clicking the field which has the aggregate function in the Layer properties > attribute forms tab leads to a slow response in opening the attribute form every single time.

Some testing showed this behaviour was related to the actual polygon layer the aggregate function was referencing. When I simplified the geometry of that polygon layer, the slow behaviour was gone (or not noticeable).

For me, I found a different expression that gave the same needed result without slowing down QGIS:
array_to_string(overlay_within('polygon_layer',"fieldname"))

This expression also populates the field in a point layer with a value from a specific field of the polygon layer the point is located in.

For me, this started with this post where the aggregate function is showed
https://spatialthoughts.com/2019/04/12/summary-aggregation-qgis/
And this post pointed me to the overlay_within function:
https://gis.stackexchange.com/questions/152257/refering-to-another-layer-in-the-field-calculator-in-qgis/314573#314573

Hope it helps.
